slackclient/slackrequest.py

```diff
@@ -94,12 +104,89 @@ def do(self, token, request="?", post_data=None, domain="slack.com", timeout=Non
             if isinstance(v, (list, dict)):
                 post_data[k] = json.dumps(v)
 
+        request_args = {
+            'api_method': api_method,
+            'post_data': post_data,
+        }
+        return self.__post_http_request(token, request_args, files, timeout)
+
+    def __post_http_request(self, token, request_args, files=None, timeout=None):
+        # Check the access token's expiration timestamp before submitting
+        # the API request and refresh if expired
+        if(self.refresh_token and self.access_token_expires_at):
+            current_ts = int(datetime.now().strftime("%s")) * 1000
+            if(current_ts > self.access_token_expires_at):
+                self.access_token_expires_at = None
+                self.refresh_access_token()
+
+        # This allows us to override the default token (self.access_token) for refresh requests
+        if(self.access_token and token is None):
+            token = self.access_token
+
+        # Set user-agent and auth headers
+        headers = {
+            'user-agent': self.get_user_agent(),
+            'Authorization': 'Bearer {}'.format(token)
+        }
+
         # Submit the request
-        return requests.post(
-            url,
+        res = requests.post(
+            'https://{0}/api/{1}'.format(self.domain, request_args['api_method']),
             headers=headers,
-            data=post_data,
+            data=request_args['post_data'],
             files=files,
             timeout=timeout,
             proxies=self.proxies
         )
+        response_json = res.json()
+        # if the API request returns an invalid_auth error, refresh the token and try again
+        if (res.status_code is 200 and response_json['ok'] is False):
+            if self.access_token is None and 'refresh_token' not in request_args['post_data']:
+                # If the API returns 'ok' false, attempt to refresh the client's access token
+                self.refresh_access_token()
+                # If token refresh was successful, retry the original API request
+                return self.__post_http_request(
+                    self.access_token,
+                    request_args
+                )
+        return res
+
+    def refresh_access_token(self):
+        """
+        Refresh the client's OAUth access tokens
+        https://api.slack.com/docs/rotating-and-refreshing-credentials
+        """
+        request_args = {
+            'api_method': 'oauth.access',
+            'post_data': {
+                'token': self.refresh_token,
```

Comment:
`token` is not a useful argument for this API call.

Reply:
True, for some reason I thought that was required.